glean-core/csharp/Glean/Glean.cs

```diff
@@ -334,13 +336,55 @@ internal void TestDestroyGleanHandle()
-            GleanInternalMetrics.architecture.SetSync("test");
-            Log.Information("Check has value");
-            bool hasValue = GleanInternalMetrics.architecture.TestHasValue();
-            Log.Information($"Has value {hasValue}");
-            string storedvalue = GleanInternalMetrics.architecture.TestGetValue();
-            Log.Information($"InitializeCoreMetrics - has value {hasValue} and that's {storedvalue}");
+            // The `Environment.OSVersion` object will return a version that represents an approximate
+            // version of the OS. As the MSDN docs state, this is unreliable:
+            // https://docs.microsoft.com/en-us/dotnet/api/system.environment.osversion?redirectedfrom=MSDN&view=netstandard-2.0
+            // However, there's really nothing we could do about it. Unless the product using the
+            // Glean SDK correctly marks their manifest as Windows 10 compatible, this API will report
+            // Windows 8 version (6.2). See the remarks section here:
+            // https://docs.microsoft.com/en-gb/windows/win32/api/winnt/ns-winnt-osversioninfoexa?redirectedfrom=MSDN#remarks
+            try
+            {
+                GleanInternalMetrics.osVersion.SetSync(Environment.OSVersion.Version.ToString());
+            }
+            catch (InvalidOperationException)
+            {
+                GleanInternalMetrics.osVersion.SetSync("Unknown");
+            }
+
+            // Possible values for `RuntimeInformation.ProcessArchitecture` are documented here:
+            // https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.architecture?view=netstandard-2.0
+            GleanInternalMetrics.architecture.SetSync(RuntimeInformation.ProcessArchitecture.ToString());
+
+            try
+            {
+                // CurrentUiCulture is used for representing the locale of the UI, coming from the OS,
+                // while CurrentCulture is the general locale used for other things (e.g. currency).
+                GleanInternalMetrics.locale.SetSync(CultureInfo.CurrentUICulture.Name);
+            }
+            catch (Exception)
+            {
+                GleanInternalMetrics.locale.SetSync("und");
+            }
+
+            if (configuration.channel != null)
+            {
+                GleanInternalMetrics.appChannel.SetSync(configuration.channel);
+            }
+
+            // Try to get the version of the product using the Glean SDK. Unfortunately,
+            // this uses reflection.
+            var mainAssembly = System.Reflection.Assembly.GetEntryAssembly();
+            if (mainAssembly != null)
+            {
+                GleanInternalMetrics.appDisplayVersion.SetSync(mainAssembly.GetName().Version.ToString());
+            }
+            else
+            {
+                GleanInternalMetrics.appDisplayVersion.SetSync("Unknown");
+            }
+
+            GleanInternalMetrics.appBuild.SetSync(configuration.buildId ?? "Unknown");
```

Comment:
In Python, we were using `_application_id` and `_application_version`
```
 metrics.glean.internal.metrics.app_build.set(cls._application_id)
 if cls._application_version is not None:
      metrics.glean.internal.metrics.app_display_version.set(
         cls._application_version)
```
I suppose we might be able to consider to adopt {`applicationId, applicationVersion`} from `Glean Initialize()`

Reply:
In Python we're doing the wrong thing: it's a bug, we're setting the wrong value for `app_build` (see [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1647782)). Moreover, in Python I'm not sure we can determine the application version easily, while this is possible in C#. We're currently providing the `buildId` from the configuration here, which is something that we won't be able to infer automatically.